### ISSUE TITLE
oidc provider assumes username or email claim exists and is stable

### ISSUE THREAD

**Issue by rhansen** [vouch/vouch-proxy#309]
When `oauth.provider` is set to `oidc`, Vouch assumes that either `username` or `email` exists in the UserInfo response. If neither exists (as is the case with GitLab when `scope=openid`), `structs.User.Username` [is the empty string](https://github.com/vouch/vouch-proxy/blob/7f2d675241d14fe6e60749ee16dab44f89af6a2c/pkg/structs/structs.go#L40-L45) and the `/validate` endpoint [fails with "no User found in jwt"](https://github.com/vouch/vouch-proxy/blob/7f2d675241d14fe6e60749ee16dab44f89af6a2c/handlers/validate.go#L49).

Furthermore, the `username` (or `email`) claim appears to be used as a unique identifier for the user. This is forbidden by [section 5.7 of the core spec](https://openid.net/specs/openid-connect-core-1_0.html#ClaimStability): "The `sub` (subject) and `iss` (issuer) Claims, used together, are the only Claims that an RP can rely upon as a stable identifier for the End-User, [...] other Claims such as `email`, `phone_number`, and `preferred_username` MUST NOT be used as unique identifiers for the End-User."

**Comment by bnfinet** [comment]
@rhansen Thanks for opening this issue.

There has been talk from time to time of moving to `sub` and it's clearly the right choice.

I know you've tested the PR against `gitlab`, have you had an opportunity to test any other providers?

**Comment by rhansen** [comment]
> I know you've tested the PR against `gitlab`, have you had an opportunity to test any other providers?

No, I have not.

**Comment by bnfinet** [comment]
@rhansen sorry for the delay in reviewing this PR

Unfortunately #310 breaks `github` support.  All of the `GetUserInfo` calls should be audited for each provider to ensure that `User.Sub` is populated.  Are you able to support such an effort?

**Comment by bnfinet** [comment]
@rhansen I've added the Sub adjustment for `github` and a rudimentary fix in `structs.PrepareUserData` but it should still be looked at a bit closer IMHO.

**Comment by UberKitten** [comment]
I noticed this issue today when playing around with authenticating guest users in Azure AD through OIDC. In AAD guest users are not in the directory themselves, just authorized to log in to the directory with an external email. The user info has sub and email but no username so it fails.

**Comment by bnfinet** [comment]
@UberKitten that sounds like #314

**Comment by UberKitten** [comment]
@bnfinet Ah yes, it is. I've been getting the "no User found in jwt" error but that's the root cause. Thanks!

**Comment by rissson** [comment]
I ran against this problem too. I need this fix, because I need to have multiple domains in `vouch.domains`. Unfortunately not all my users have an email address in those domains and thus get denied login.

**Comment by bnfinet** [comment]
@rissson have you tried 'allowAllUsers' config item?

**Comment by rissson** [comment]
Yes, that works, but then I can't use vouch proxy under different domains.

**Comment by bnfinet** [comment]
@rissson If your users are not all within the same domain you may need to run multiple VPs, one for each domain, each with `allowAllUsers` and its own callback URL. 

 `domains` is meant for orgs in which all users have email addresses within the specified domain.

**Comment by bnfinet** [comment]
@rhansen are you still interested in supporting PR #310 ?  I'd love to work with you to get it merged.

**Comment by djcrabhat** [comment]
We designed a custom OIDC system where that's all UserInfo is giving back (and checking token's validity).  Per https://openid.net/specs/openid-connect-core-1_0.html#UserInfoResponse

> The sub (subject) Claim MUST always be returned in the UserInfo Response.

VouchCookie gets set, when /validate is called, it's errors "no User found in jwt", because the Username field is blank https://github.com/vouch/vouch-proxy/blob/master/handlers/validate.go#L48
Would love if we could get this to work.  I know moving Sub up to the common user object is breaking gitlab.  Can we make a new custom OidcUser struct, add `Sub` to that?  Then a `PrepareUserData` like
```
func (u *OidcUser) PrepareUserData() {
	if u.Username == "" {
	        if u.email == "" {
                   u.Username = u.Sub
		} else {
		    u.Username = u.Email
               }
	}
}
```

that way, at least /validate has _something_ to fall back on.

**Comment by bnfinet** [comment]
@djcrabhat thanks for the kind words and for taking a look at this.

Have you tested #310 against your OIDC setup?  Does it work?

from Dec 6th to @rhansen ..
> Unfortunately #310 breaks github support. All of the GetUserInfo calls should be audited for each provider to ensure that User.Sub is populated. Are you able to support such an effort?

Would you be in a position to perform such an audit and/or add tests for such?

**Comment by djcrabhat** [comment]
@bnfinet will compile and test locally tomorrow (hopefully).  Once I get that going I'd be happy to take a whack at getting it well exercised in the tests.

**Comment by bnfinet** [comment]
@plachor no I can't provide you with an estimate at this time.  Thanks or you patience.

Everything else you say seems correct.

**Comment by plachor** [comment]
Thnx for clarification.

**Comment by rhansen** [comment]
@bnfinet I'm not using vouch anymore and I don't really have the time to invest in fixing this issue. I rebased PR #310 onto latest `master` and cleaned up the commits bit in case someone wants to pick up where I left off. The changes are untested (other than GitHub actions).

**Comment by ShyLionTjmn** [comment]
I have allowAllUsers: true, but still, user gets redirect loop, if it does not have email.
